lib/horde/horde_dynamic_supervisor.ex

```diff
@@ -881,9 +1104,28 @@ defmodule Horde.DynamicSupervisor do
     end
   end
 
+  defp handle_graceful_shutdown_horde({child_id, start, restart, shutdown, type, modules}, state) do
+    if state.graceful_shutdown_manager do
+      child_spec = %{
+        id: child_id,
+        start: start,
+        restart: restart,
+        shutdown: shutdown,
+        type: type,
+        modules: modules
+      }
+
+      GenServer.cast(state.graceful_shutdown_manager, {:shut_down, child_spec})
+    end
+  end
+
+  defp handle_graceful_shutdown_horde(_, _) do
```

Comment:
what other cases are we handling that this will match?

Reply:
On line 859 it protects against the child spec being nil for a given pid.

Reply:
I am still wondering about this line. I am also wondering whether line 1071 should be removed (and whether this will solve the issue of the missing child specs).

The way DynamicSupervisor terminates a pid is by monitoring it, sending a `:shutdown` signal, and then processing the `:DOWN` tuples as they come in. I guess a `:DOWN` tuple could come out of order (eg, spuriously), which is why there is still a catch-all match on 1070. But as far as I can tell, we should *always* come out at either 1065 or 1075 for every pid that we are terminating.

Thoughts?

Reply:
After going through the code and tracing it in my head I don't think there is a case for it. I removed it, I tend to like to add catch all things to avoid crashing the supervisor, but if something like that happens "let it crash".